### ISSUE TITLE
WASM: Fused backprop inner loop to eliminate TS/WASM boundary crossings

### ISSUE THREAD

**Issue by stservice** [stSoftwareAU/NEAT-TS#1520]
## Summary

Create a fused WASM function that performs the complete backpropagation inner loop for a single neuron - combining derivative calculation, error distribution, weight accumulation, and bias accumulation into a single WASM call. Currently, backpropagation makes multiple separate WASM calls per neuron (wasmSquash, wasmUnSquash, wasmCalculateError, derivative batches) plus TS arithmetic for weight/bias updates. Each boundary crossing has ~8.7ns overhead that compounds across thousands of neurons.

## What Needs to Be Done

1. **Create benchmark** measuring full backprop pass time on large creatures, profiling time spent in WASM boundary crossings vs computation
2. **Design fused Rust function** `fused_backprop_neuron` that takes:
   - Neuron's inbound synapse weights and source activations
   - Target error/adjustment for the neuron
   - Learning rate and training config parameters
   - Returns: updated weights, updated bias, distributed errors for source neurons
3. **Implement in Rust** combining:
   - Derivative calculation (already in WASM)
   - Error calculation (already in WASM)
   - Weight accumulation (from #1518)
   - Bias accumulation (from #1518)
4. **Keep all intermediate values in WASM memory** - no round-trip to JS between steps
5. **Add TypeScript integration** that prepares flat arrays for the fused call
6. **Run benchmark** comparing fused vs current multi-call approach

## Expected Performance Gains

- **Estimated 3-5x speedup on the backprop inner loop** due to:
  - Elimination of N boundary crossings per neuron (currently ~5-8 WASM calls per neuron)
  - Intermediate values stay in WASM registers/stack (no JS heap allocation)
  - Better instruction cache utilisation (single function vs multiple dispatches)
  - SIMD across all operations in the fused pipeline
- This is the **highest-impact single optimisation** for training performance

## Effort Estimate

- **High** (~4-5 days): Requires combining multiple existing WASM functions into a coherent pipeline. Careful attention to data layout and error propagation semantics. Benefits from #1518 and #1519 being completed first.

## Acceptance Criteria

- [ ] Before/after benchmarks on large creatures (500+ neurons, 2000+ synapses) showing measurable training speedup
- [ ] All existing backpropagation tests pass unchanged
- [ ] Training convergence behaviour is numerically equivalent (within floating-point tolerance)
- [ ] External API unchanged
- [ ] Australian English in all code/comments
- [ ] Graceful fallback to current multi-call approach if WASM unavailable

## Dependencies

- Benefits from #1518 (weight/bias accumulation in WASM) and #1519 (elastic distribution in WASM), but can be developed independently using already-existing WASM primitives (derivative, error calculation)

## Context

Part of #1517

Benchmark data from `bench/BackpropWasmOverhead.ts` shows individual WASM calls have ~8.7ns overhead. For a neuron with 50 inbound connections making 5-8 WASM calls, that's ~40-70ns of pure overhead per neuron per backprop iteration. On a 500-neuron network trained over 1000 iterations, this is ~20-35ms of pure boundary-crossing overhead - a significant fraction of total training time.

**Comment by stservice** [comment]
## Automated Processing Failed (First Attempt)

The automated worker was unable to complete this issue on the first attempt.

### Failure Details
Git push failed while pushing branch `issue-1520-wasm-fused-backprop-inner-loop-to-eliminate-ts-was` to origin.

This typically happens due to:
- Authentication/permission issues
- Protected branch rules
- Network connectivity problems
- Missing `workflow` scope for OAuth tokens (when modifying GitHub Actions files)

<details>
<summary>Git push error output (click to expand)</summary>

```
To https://github.com/stSoftwareAU/NEAT-AI.git
 ! [rejected]          issue-1520-wasm-fused-backprop-inner-loop-to-eliminate-ts-was -> issue-1520-wasm-fused-backprop-inner-loop-to-eliminate-ts-was (fetch first)
error: failed to push some refs to 'https://github.com/stSoftwareAU/NEAT-AI.git'
hint: Updates were rejected because the remote contains work that you do not
hint: have locally. This is usually caused by another repository pushing to
hint: the same ref. If you want to integrate the remote changes, use
hint: 'git pull' before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
</details>

### What happens next?
- This issue has been marked with the `failed-once` label
- The worker will **automatically retry** this issue on the next scan
- If it fails again, the issue will be marked as `failed` and require manual intervention

### To prevent automatic retry
If you want to work on this manually, remove the `failed-once` label and assign yourself.

**Comment by stservice** [comment]
## Negative Result: Fused WASM Backprop Inner Loop

After full implementation and benchmarking, the fused WASM approach is **3-5x slower** than the current multi-call TypeScript approach. Issue closed as wontfix.

### Benchmark Results (Apple M4 Pro, Deno 2.6.8)

| Synapse Count | Multi-call TS (baseline) | Fused WASM | Ratio |
|---|---|---|---|
| 10 | 84.1 ns | 449.5 ns | 5.34x slower |
| 50 | 282.9 ns | 1.1 µs | 3.96x slower |
| 200 | 1.0 µs | 3.1 µs | 3.10x slower |

### Root Cause

The issue's premise was based on **outdated assumptions**:

1. **Issue #1377 already eliminated most boundary crossings.** The `fusedErrorDistribution` from Issue #1377 combines `calculateError + safeZoneAdjustment + elastic distribution` into a single WASM call. After that optimisation, only 2-3 scalar WASM calls remain per neuron (`wasmUnSquash`, `wasmSquash`), totalling ~17ns of overhead.

2. **Array marshalling overhead exceeds savings.** The fused function requires packing/unpacking `SynapseState` data (8 f64 values × N synapses) into `Float64Array` buffers. This costs ~400-1000ns, far exceeding the ~17ns saved.

3. **Weight/bias accumulation is efficient in TypeScript.** Simple arithmetic that V8 JIT-compiles very well. No computational benefit from WASM.

### What Was Built

- Rust `fused_backprop_neuron` function combining weight accumulation + bias accumulation + unsquash + squash
- TypeScript bridge and integration into `Neuron.propagate()`
- 181 Rust tests + 8 TypeScript tests passing
- All 822 existing propagate tests passing

### Recommendation

Further backprop performance gains should target:
- Reducing computational work (sparse updates, batch processing across samples)
- Not eliminating the remaining 2-3 scalar WASM calls which have negligible overhead

A negative result is still a valuable result.

**Comment by stservice** [comment]
## Automated Processing Failed (Second Attempt - Permanently Failed)

The automated worker has failed to complete this issue twice and will not retry automatically.

### Failure Details
Claude completed without making any changes. This may indicate the issue is too complex, too vague, or Claude entered plan mode without implementing.

### What happened?
- This issue failed on the first attempt and was marked with `failed-once`
- The worker retried and failed again on the second attempt
- The issue has been marked with the `failed` label to prevent repeated attempts
- Manual intervention is required

### Why this might be happening
- The issue may be too complex or large for automated processing
- The issue description may need more detail or clarification
- There may be dependencies or permissions that Claude doesn't have

### To retry this issue
1. Investigate and address the underlying problem (check the failure details above)
2. Consider simplifying or breaking down the issue into smaller tasks
3. Remove the `failed` label from this issue
4. The worker will pick it up again on the next scan
